File: Orimath/ViewModels/GlobalSettingEffect.cs
using System;
using System.IO;
using Orimath.IO;
using Orimath.Plugins;
using ApplicativeProperty;

namespace Orimath.ViewModels
{
    public class GlobalSettingEffect : IParametricEffect
    {
        private GlobalSetting _rootSetting;
        private GlobalSetting? _setting;

        public object GetParameter() => _setting = (GlobalSetting)_rootSetting.Clone();

        public string[] MenuHieralchy => new[] { "設定" };

        public string Name => "環境設定";

        public string ShortcutKey => "";

        public Stream? Icon => null;

        public GlobalSettingEffect(GlobalSetting setting)
        {
            _rootSetting = setting;
        }

        public IGetProp<bool> CanExecute => Prop.True;

        public void Execute()
        {
            if (_setting is null) return;

            _rootSetting.ViewSize = _setting.ViewSize;
            Settings.Save(SettingName.Global, _rootSetting);
        }
    }
}

File: Orimath/ViewModels/GlobalSettingEffect.fs
namespace Orimath.ViewModels
open Orimath.IO
open Orimath.Plugins
open ApplicativeProperty
open Orimath.Internal

type GlobalSettingEffect(rootSetting: GlobalSetting) =
    let mutable setting = Unchecked.defaultof<_>

    interface IParametricEffect with
        member val MenuHieralchy = [| "設定" |]
        member _.Name = "環境設定"
        member _.ShortcutKey = ""
        member _.Icon = null
        member _.GetParameter() =
            setting <- rootSetting.Clone()
            box setting
        member _.CanExecute = upcast Prop.ctrue
        member _.Execute() =
            if isNotNull (box setting) then
                rootSetting.ViewSize <- setting.ViewSize
                Settings.save Settings.Global rootSetting
